### DISCUSSION TITLE
Running ReactPHP in DDEV

### DISCUSSION THREAD

**Discussion by BernhardBaumrock** [ddev/ddev#5973] [General]
Hi DDEVs!

TL;DR

How can I expose 127.0.0.1:1234 from inside docker to myservice.myapp.ddev.site on the host?

---

Long Story:

I'm evaluating https://reactphp.org/ for my next project and I'd like to properly run it with DDEV right from the start :) I know that DDEV can run additional services, but I got stuck on the docs early in the process...

What works so far:

I created a new ddev project, logged into the container via `ddev ssh` and there I ran `composer require react/react:^1.4`

Then I created `server.php`

```php
<?php

// $ composer require react/http react/socket # install example using Composer
// $ php example.php # run example on command line, requires no additional web server

require __DIR__ . '/vendor/autoload.php';

$server = new React\Http\HttpServer(function (Psr\Http\Message\ServerRequestInterface $request) {
  return React\Http\Message\Response::plaintext(
    "Hello World!\n"
  );
});

$socket = new React\Socket\SocketServer('127.0.0.1:8080');
$server->listen($socket);

echo "Server running at http://127.0.0.1:8080" . PHP_EOL;
```

Now I ran `php server.php` and it showed `Server running at http://127.0.0.1:8080` - great!

But the problem is: How can I expose that server/service from my DDEV to the host?

I tried to run `php server.php` on my host and then I can go to http://127.0.0.1:8080 in the browser and I see `Hello World!` - great :)

What I'd like to achieve is that this server is started with the project and I can reach it from my web app from a custom url, something like myservice.reactphp.ddev.site; I read that there is support for supervisord so that should be possible I guess, but first I need to get it running manually...

I tried to follow https://ddev.readthedocs.io/en/stable/users/extend/custom-compose-files/ and created the file `.ddev/docker-compose.ports.yaml` with the following content:

```yaml
services:
  someservice:
    ports:
      - "9999:9999"
```

But I got this error:

Failed to restart reactphp: composeCmd failed to run 'COMPOSE_PROJECT_NAME=ddev-reactphp docker-compose -f /Users/bernhard/baumrock/reactphp/.ddev/.ddev-docker-compose-base.yaml -f /Users/bernhard/baumrock/reactphp/.ddev/docker-compose.ports.yaml config', action='[]', err='exit status 15', stdout='', stderr='service "someservice" has neither an image nor a build context specified: invalid compose project' 

I'm a little lost as I'm a little blank when it comes to docker or networking in general.

Thx in advance for your help!

**Comment by rfay** [comment]
Probably you want to use `web_extra_daemons` and `web_extra_exposed_ports`, see https://ddev.readthedocs.io/en/stable/users/extend/customization-extendibility/#running-extra-daemons-in-the-web-container

Remember that you'll have to configure react-php to listen on 0.0.0.0, not 127.0.0.1, which is available only from inside the container.

**Reply by BernhardBaumrock** [reply]
Nice, thx @rfay I'll have a look!

**Reply by BernhardBaumrock** [reply]
Hey @rfay unfortunately I get a "bad gateway" when I try to visit https://reactphp.ddev.site:3000

My react server seems to be running fine as when I'm doing `curl http://127.0.0.1:3000` from inside the container I get `Hello World!`

This is my config.yaml

```yaml
name: reactphp
type: php
docroot: ""
php_version: "8.2"
webserver_type: apache-fpm
xdebug_enabled: false
additional_fqdns: []
database:
  type: mariadb
  version: "10.6"
omit_containers: [dba]
use_dns_when_possible: true
timezone: Europe/Vienna
composer_version: "2"

web_extra_exposed_ports:
  - name: reactphp
    container_port: 3000
    http_port: 2999
    https_port: 3000

web_extra_daemons:
  - name: "reactphp"
    command: "php /var/www/html/server.php"
    directory: /var/www/html
```

BTW: I didn't find info in the docs about what the "directory" in the last line does and whether it is necessary or not, I guess in my case it's obsolete as I'm starting the server.php from an absolute path?

Thx for your help!

**Reply by stasadev** [reply]
Hi @BernhardBaumrock,

Did you change the address in the `server.php`?

```diff
-$socket = new React\Socket\SocketServer('127.0.0.1:8080');
+$socket = new React\Socket\SocketServer('0.0.0.0:3000');
```

**Reply by BernhardBaumrock** [reply]
NICE!!! It works :) I knew it must be simple if you know what to do. Randy even wrote that yesterday, but I totally missed that part as I was in a hurry. Thx for your help!

**Reply by rfay** [reply]
Added 
* https://github.com/ddev/ddev/pull/5981

**Reply by BernhardBaumrock** [reply]
Cool, thx! :)
